### PR TITLE
NOISSUE: remove any rest api call from kube api flow

### PR DESCRIPTION
**Pull request by RazRegev**
replace any rest api call to edit resources in assisted service
with matching kube-api call instead.

### CODE REVIEW

File: discovery-infra/start_discovery.py [hunk 1]
```diff
@@ -500,6 +540,42 @@ def set_hosts_roles(client, cluster, nodes_details, machine_net, tf, master_coun
...
+    )
+    return utils.get_libvirt_nodes_from_tf_state(
+        network_names=networks_names,
+        tf_state=tf.get_state(),
+    )
+
+
+def set_agent_hostname(agent, nodes_details, is_ipv4):
+    if is_ipv4 and not args.with_static_network_config:
+        return
```
Comment (osherdp) [inline discovery-infra/start_discovery.py:558]: append blank line

**Comment by osherdp** [comment]
@RazRegev maybe retitle the PR?

File: discovery-infra/start_discovery.py [hunk 2]
```diff
@@ -409,28 +413,51 @@ def nodes_flow(client, cluster_name, cluster, machine_net, kube_client=None, clu
...
                     client, cluster.id, main_cidr, ipv4_first=is_ipv4)})
-                set_cluster_machine_cidr(client, cluster.id, machine_net, set_vip_dhcp_allocation=False)
+                if args.kube_api:
+                    cluster_deployment.patch(
+                        install_strategy=InstallStrategy(
+                            host_prefix=args.host_prefix if is_ipv4 else args.host_prefix6,
+                            machine_cidr=get_machine_cidr_from_machine_net(machine_net),
+                            cluster_cidr=args.cluster_network if is_ipv4 else args.cluster_network6,
+                            service_cidr=args.service_network if is_ipv4 else args.service_network6,
+                            ssh_public_key=args.ssh_key,
```
Comment (osherdp) [inline discovery-infra/start_discovery.py:449]: I think @nmagnezi moved this field to ``InstallEnv``, no?

File: discovery-infra/start_discovery.py [hunk 3]
```diff
@@ -409,28 +413,51 @@ def nodes_flow(client, cluster_name, cluster, machine_net, kube_client=None, clu
...
+                            control_plane_agents=args.master_count,
+                            worker_agents=args.number_of_workers,
+                        ),
+                    )
```
Comment (osherdp) [inline discovery-infra/start_discovery.py:453]: Do we even need this patch?
I thought we already set all those fields with the right value
Alternatively, we must re-specify all the fields here?

File: discovery-infra/start_discovery.py [hunk 4]
```diff
@@ -500,6 +540,42 @@ def set_hosts_roles(client, cluster, nodes_details, machine_net, tf, master_coun
...
+        tf_vars['libvirt_network_name'],
+        tf_vars['libvirt_secondary_network_name'],
+    )
+    return utils.get_libvirt_nodes_from_tf_state(
+        network_names=networks_names,
+        tf_state=tf.get_state(),
+    )
+
+
+def set_agent_hostname(agent, nodes_details, is_ipv4):
```
Comment (osherdp) [inline discovery-infra/start_discovery.py:556]: maybe some of the functions should be in some utility module for kubeapi functions

**Review by osherdp — CHANGES_REQUESTED** [review]

File: discovery-infra/start_discovery.py (hunk 3 above)
Comment (RazRegev) [reply discovery-infra/start_discovery.py]: you're right, this is unnecessary

File: discovery-infra/test_infra/kubeapi_utils.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        functools.partial(get_hostname_for_agent, nodes_details, agent),
+        timeout_seconds=60,
+        sleep_seconds=1,
+        waiting_for=f'agent={agent.ref} to find a hostname',
+    )
+    logger.info('patching agent hostname=%s', hostname)
+    agent.patch(hostname=hostname)
+
+
+def get_hostname_for_agent(nodes_details, agent):
```
Comment (osherdp) [inline discovery-infra/test_infra/kubeapi_utils.py:84]: It bugs me that the order of the parameters is not the same as above function

File: discovery-infra/test_infra/kubeapi_utils.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    def get_bmc_address():
+        interfaces = agent.status().get('inventory', {}).get('interfaces')
+        if not interfaces:
+            return
+        ip_addresses = interfaces[0].get(
+            'ipV4Addresses' if is_ipv4 else 'ipV6Addresses'
+        )
+        if not ip_addresses:
+            return
```
Comment (osherdp) [inline discovery-infra/test_infra/kubeapi_utils.py:27]: append blank line

File: discovery-infra/test_infra/kubeapi_utils.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        interface_class = IPv4Interface if is_ipv4 else IPv6Interface
+        return str(interface_class(ip_addresses[0]).ip)
```
Comment (osherdp) [inline discovery-infra/test_infra/kubeapi_utils.py:29]: not sure it's much Pythonic all this area
can we change it in some way?

File: discovery-infra/install_cluster.py [hunk 8]
```diff
@@ -85,60 +85,131 @@ def wait_till_installed(client, cluster, timeout=60 * 60 * 2):
 # 2. Running install cluster api
 # 3. Waiting till all nodes are in installing status
 # 4. Downloads kubeconfig for future usage
-def run_install_flow(client, cluster_id, kubeconfig_path, pull_secret,
-                     tf=None, kube_client=None, cluster_deployment=None):
+def run_install_flow(
```
Comment (osherdp) [inline discovery-infra/install_cluster.py:88]: yes, that's a much better approach :)
Maybe in the future we'll have two classes that implements the same interface, one for REST and one for kube-api

**Review by osherdp — CHANGES_REQUESTED** [review]

File: discovery-infra/test_infra/kubeapi_utils.py (hunk 5 above)
Comment (RazRegev) [reply discovery-infra/test_infra/kubeapi_utils.py]: bless you

File: discovery-infra/install_cluster.py (hunk 8 above)
Comment (RazRegev) [reply discovery-infra/install_cluster.py:88]: Requires a huge refactoring but I couldn't agree more

**Review by osherdp — APPROVED** [review]

**Pull request merged**